### PR TITLE
eth_sendRawTransaction JSON RPC 

### LINKED ISSUE #1071: Feature: Sending signed transaction to network via RPC.
Is there any possibility to send an already signed transaction to the network via geth/RPC?

This would be a good security feature because it would allow offline signing of transactions like it is possible with Armory with Bitcoins.

### PR DESCRIPTION
**Pull request by SilentCicero**
I'm still testing this, but this should allow signed transactions to be sent to JSON RPC. It uses the pushTx method, which I believe is the right approach for sending signed transaction packets to the chain. Thoughts?

Closes #1071

### CODE REVIEW

**Comment by d11e9** [comment]
:+1: handy

**Comment by tgerring** [comment]
I definitely would like to see this get merged. Once concern I have is that with this in place, it seem like eth_sendTransaction should then be using the same underlying XEth signing and pushing functions.

For example, we have XEth.Sign() and XEth.PushTx(), but neither of these are used by XEth.SendTransaction(), which seems a bit strange. Maybe we leave refactoring for a later date, since there's probably quite a bit of cleanup that can be done, but I wanted to document this concern at a minimum.

File: rpc/api.go [hunk 1]
```diff
@@ -170,6 +170,17 @@ func (api *EthereumApi) GetRequestReply(req *RpcRequest, reply *interface{}) err
 		}
 		*reply = v
 
+	case "eth_pushTx":
```
Comment (obscuren) [inline rpc/api.go]: Please rename this to `eth_sendRawTransaction`

Comment (SilentCicero) [inline rpc/api.go]: I felt it in my bones that you wanted this name :)

**Comment by SilentCicero** [comment]
@tgerring I think eth_sendRawTransaction should remain an open method outside the account lock mechanism that allows for valid and complete signed TX's to stream into RPC. I see there being many use cases for this, I need it for several projects myself. In general browser-based account management. I'm working with Martin and Aaron on this. Cool stuff ahead!

**Comment by kumavis** [comment]
Definitely want `eth_sendRawTransaction`. Haven't reviewed the implementation here though.

**Comment by ethers** [comment]
Please add test/s too.  maybe in https://github.com/ethereum/rpc-tests

File: rpc/api/utils.go [hunk 2]
```diff
@@ -51,6 +51,7 @@ var (
 			"getData",
 			"getCode",
 			"sign",
+			"eth_sendRawTransaction",
```
Comment (subtly) [inline rpc/api/utils.go]: This one looks different than all the others.

**Comment by subtly** [comment]
Not quite groking what feedback the errors provide. What if there's an error? Does this PR check the nonce, verify transaction will run, etc?

Would be great to see this go in, and even better if the feedback is as informative as submitting a transaction without a signature.

**Comment by SilentCicero** [comment]
Just testing with ethereumjs-tx, and curl I get successful results back. All invalid TX data reported back successfully through RPC errors, i.e. low gas, invalid nonce, invalid from data. It acted much like Transaction or eth_sendTransaction. Although more tests should be run to fully vet PushTx. 

My curl sample is like this:
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_sendRawTransaction","params":["0f6cf08300b7bdc6d83d1898321a19770cb25578","0xf869808609184e72a000831cfde094db7f3f1863c261524fa069127edc7fa9d568359683015f90001ba08a0ea4c989a033c2b41128b11cdc7224303eddb9965e30c2403e3037eb4be8cca0fb199be30641029857c4ef19fea6f0c444957a0a9e6110eb9a005d498a8dc82b"],"id":67}' http://localhost:8545

Where unfortunately, the first param used here, must be a "from" address. @tgerring this is unnecessary, but I feel uncomfortable manipulating core to remove this required "from", at this time. And seeing as the "from" param is required in both new and old style transactions, leaving it in seems entirely adequate.

The address used here is:
0f6cf08300b7bdc6d83d1898321a19770cb25578

The private key used here is:
904c942217570531bcf3e51f76686132904c942217570531bcf3e51f76686132

The raw signed transaction used here is:
0xf869808609184e72a000831cfde094db7f3f1863c261524fa069127edc7fa9d568359683015f90001ba08a0ea4c989a033c2b41128b11cdc7224303eddb9965e30c2403e3037eb4be8cca0fb199be30641029857c4ef19fea6f0c444957a0a9e6110eb9a005d498a8dc82b

The library used to generate this transaction data can be found here: https://github.com/ethereum/ethereumjs-tx

Cheers :)

**Comment by ethers** [comment]
This really should have tests in code which will help reviewers and future refactoring and maintenance of this code.  Anyway, I won't mention again and will let the reviewers decide between the balance of quickly merging code/functionality vs time to write tests (sometimes the tests could take much longer to write than the code).

> I feel uncomfortable manipulating core to remove this required "from", at this time.

Is it possible to have the first param still be an object with a `from` key ?
```
"params":[{"from":"0f6cf08300b7bdc6d83d1898321a19770cb25578", "data":"0xf869808609184e72a000831cfde094db7f3f1863c261524fa069127edc7fa9d568359683015f90001ba08a0ea4c989a033c2b41128b11cdc7224303eddb9965e30c2403e3037eb4be8cca0fb199be30641029857c4ef19fea6f0c444957a0a9e6110eb9a005d498a8dc82b"}]
```

**Comment by SilentCicero** [comment]
@ethers with the latest commit, thanks to the help of @tgerring you no longer need to add the "from" param inside the RPC "params", example:

```curl -X POST --data '{"jsonrpc":"2.0","method":"eth_sendRawTransaction","params":["0xf869808609184e72a000831cfde094db7f3f1863c261524fa069127edc7fa9d568359683015f90001ba08a0ea4c989a033c2b41128b11cdc7224303eddb9965e30c2403e3037eb4be8cca0fb199be30641029857c4ef19fea6f0c444957a0a9e6110eb9a005d498a8dc82b"],"id":67}' http://localhost:8545```

The address used:
```0f6cf08300b7bdc6d83d1898321a19770cb25578```

The private key used:
```904c942217570531bcf3e51f76686132904c942217570531bcf3e51f76686132```

I don't feel comfortable writing tests for this in go at this point, however, you can do a run through with CURL using the CURL example I provided above. Set the address above as your --etherbase="0f6cf08300b7bdc6d83d1898321a19770cb25578" mine ether to it, then use the CURL sample above to run eth_sendRawTransaction.

There is definitely more testing to do here.

**Comment by SilentCicero** [comment]
We also need to catch invalid gas, gasPrice, nonce and to much like eth_sendTransaction does. Both with the RPC and with xeth.

**Comment by tgerring** [comment]
The tests were never merged... I'll create a separate PR with them

**Pull request merged**